go/vt/vtgate/engine/vexplain.go

```diff
@@ -0,0 +1,213 @@
+	"vitess.io/vitess/go/vt/srvtopo"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+type (
+	ExecuteEntry struct {
+		ID        int
+		Target    *querypb.Target
+		Gateway   srvtopo.Gateway
+		Query     string
+		FiredFrom Primitive
+	}
+
+	VExplain struct {
+		Input Primitive
+		Type  sqlparser.VExplainType
+	}
+)
+
+var _ Primitive = (*VExplain)(nil)
+
+// RouteType implements the Primitive interface
+func (v *VExplain) RouteType() string {
+	return v.Input.RouteType()
+}
+
+// GetKeyspaceName implements the Primitive interface
+func (v *VExplain) GetKeyspaceName() string {
+	return v.Input.GetKeyspaceName()
+}
+
+// GetTableName implements the Primitive interface
+func (v *VExplain) GetTableName() string {
+	return v.Input.GetTableName()
+}
+
+// GetFields implements the Primitive interface
+func (v *VExplain) GetFields(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return v.Input.GetFields(ctx, vcursor, bindVars)
+}
+
+// NeedsTransaction implements the Primitive interface
+func (v *VExplain) NeedsTransaction() bool {
+	return v.Input.NeedsTransaction()
+}
+
+// TryExecute implements the Primitive interface
+func (v *VExplain) TryExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	vcursor.Session().VExplainLogging()
+	_, err := vcursor.ExecutePrimitive(ctx, v.Input, bindVars, wantfields)
+	if err != nil {
+		return nil, err
+	}
+	return v.convertToResult(ctx, vcursor)
+}
+
+// TryStreamExecute implements the Primitive interface
+func (v *VExplain) TryStreamExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	vcursor.Session().VExplainLogging()
+	err := vcursor.StreamExecutePrimitive(ctx, v.Input, bindVars, wantfields, func(result *sqltypes.Result) error {
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	result, err := v.convertToResult(ctx, vcursor)
+	if err != nil {
+		return err
+	}
+	return callback(result)
+}
+
+func (v *VExplain) convertToResult(ctx context.Context, vcursor VCursor) (*sqltypes.Result, error) {
+	switch v.Type {
+	case sqlparser.QueriesVExplainType:
+		result := convertToVExplainQueriesResult(vcursor.Session().GetVExplainLogs())
+		return result, nil
+	case sqlparser.AllVExplainType:
+		return v.convertToVExplainAllResult(ctx, vcursor)
+	default:
+		return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "Unknown type of VExplain plan")
+	}
+}
+
+func (v *VExplain) convertToVExplainAllResult(ctx context.Context, vcursor VCursor) (*sqltypes.Result, error) {
+	logEntries := vcursor.Session().GetVExplainLogs()
+	explainResults := make(map[Primitive]string)
+	for _, entry := range logEntries {
+		if entry.Target == nil || entry.Gateway == nil || entry.FiredFrom == nil {
+			continue
+		}
+		if explainResults[entry.FiredFrom] != "" || strings.Contains(entry.Query, "where 1 != 1") {
```

Comment:
we should not filter out impossible queries here

Reply:
Done